devito/finite_differences/differentiable.py

```diff
@@ -363,12 +370,44 @@ def _eval_at(self, func):
 
 class Add(DifferentiableOp, sympy.Add):
     __sympy_class__ = sympy.Add
-    __new__ = DifferentiableOp.__new__
+
+    def __new__(cls, *args, **kwargs):
+        # Here, often we get `evaluate=False` to prevent SymPy evaluation (e.g.,
+        # when `cls==EvalDerivative`), but in all cases we at least apply a small
+        # sets of basic optimizations
+
+        # (a+b)+c -> a+b+c (flattening)
+        nested, others = split(args, lambda e: isinstance(e, Add))
+        args = flatten(e.args for e in nested) + list(others)
+
+        # a+0 -> a
+        args = [i for i in args if i != 0]
+
+        # Reorder for homogeneity with pure SymPy types
+        _addsort(args)
+
+        return super().__new__(cls, *args, **kwargs)
 
 
 class Mul(DifferentiableOp, sympy.Mul):
     __sympy_class__ = sympy.Mul
-    __new__ = DifferentiableOp.__new__
+
+    def __new__(cls, *args, **kwargs):
+        # A DifferentiableOp may not trigger evaluation upon construction
+        # (e.g., if an EvalDerivative is present among the arguments)
+        # So we treat some special cases here
+
+        # a*0 -> 0
```

Comment:
this is never an issue to the best of my knowledge

I think the only issue was with EvalDerivative (and therefore Add) because they explicitly forbid evaluation

Reply:
@mloubout turn out it was an issue in a very corner case. I added flattening + test (a test that took me a while to write out...)

Reply:
👍